### ISSUE TITLE
Can we handle multiple streams?

### ISSUE THREAD

**Issue by marcoherbst** [evercam/evercam-gateway#39]

**Comment by rajmohanb** [comment]
Will test with configuration file modifications and update about the feasibility here

**Comment by scandox** [comment]
@marcoherbst @rajmohanb Personally, I view this as a dealbreaker. If this isn't possible I would consider that we need to review how to move forward with the project.

A few notes:

*    I believe you would need a unique port for each Gstreamer stream on Janus. So we need Janus to be able to listen across a range of ports. I imagine 10 or 20 ports range would be more than enough. Obviously the Gateway code needs a way of picking a free one.
*    Ideally we would not want to restream the same camera for multiple clients. So if we can detect when a particular stream is active and relay this media to N webRTC clients that would be the best case (however if it introduces too much complexity we could live without it). However I note that at this link: https://github.com/meetecho/janus-gateway/blob/master/plugins/janus_streaming.c there is a note saying "shared streaming context for all peers attached to the stream)." which sounds like what we want.

**Comment by rajmohanb** [comment]
@scandox - I was able to stream video from 2 sources (one hikvision camera and another from a file). The only changes required were changes in the Janus config file and the UI. For any new camera source, a new stream must be added in the janus configuration file (with a different port) and then Janus must be restarted.

Two minor things we need to sort out
[1] As and when a new camera is added/removed, the same has to be added/removed from the Janus config file. And then the Janus must be restarted.

[2] The functionality of choosing a camera out of 'n' streams is with UI. This is how Janus signaling is built. As soon as a user connects to Janus, Janus sends a list of camera streams (2 for my testing) and user must chose one of them and send a "select" one od the ids. After this, Janus will start streaming from the selected video source.

I have attached the modified Janus config file and the gstreamer pipelines.

[Note]: Remove the file extension (.txt) from janus.plugin.streaming.cfg.txt, as github allows a limited file extensions to upload.

Gstreamer pipeline 1 from Camera:
gst-launch-1.0 rtspsrc location=rtsp://admin:12345@192.168.1.100/h264/ch1/main/av_stream name=src ! rtph264depay ! avdec_h264 ! videoconvert ! videoscale ! vp8enc target-bitrate=256000 error-resilient=true ! rtpvp8pay ! udpsink host=127.0.0.1 port=5004

Gstreamer pipeline 2 from local file:
gst-launch-1.0 filesrc location=Video.avi ! decodebin ! queue ! videoconvert ! videoscale ! queue ! vp8enc target-bitrate=256000 error-resilient=true ! rtpvp8pay ! udpsink host=127.0.0.1 port=5005

**Comment by rajmohanb** [comment]
I am unable to attache the cfg file as text or pdf, so pasting the configuration of 2 streams


[ IP Camera 1 ]
type = rtp
id = 1
description = Camera 1 - HikVision
audio = no
video = yes
;#audioport = 5002
;#audiopt = 111
;#audiortpmap = opus/48000/2
videoport = 5004
videopt = 100
videortpmap = VP8/90000
secret = adminpwd

[ IP Camera 2 ]
type = rtp
id = 2
description = Camera 2 - File
audio = no
video = yes
;#audioport = 5002
;#audiopt = 111
;#audiortpmap = opus/48000/2
videoport = 5005
videopt = 100
videortpmap = VP8/90000
secret = adminpwd

**Comment by rajmohanb** [comment]
Janus signaling log - `https://gist.github.com/rajmohanb/69d747c3679d045fbe9d`. In the log you can notice the 2 listed available streams and stream #2 has been chosen

```
Got a list of available streams:
[Object, Object]
  >> [1] Camera 1 - HikVision (live)
  >> [2] Camera 2 - File (live)
Selected video id #2
```

**Comment by rajmohanb** [comment]
Janus configuration file with multiple camera sources - https://gist.github.com/rajmohanb/143aa18c70034fc473a2

**Comment by scandox** [comment]
@rajmohanb It seems to me that one approach we could take is to
pre-configure say 10 video ports on Janus and then we can just use our own
code to decide which port to attach a particular stream to? The reason I
propose this, is so we can avoid having to constantly write to config files
and/or restart Janus.

Alternatively, if there is some to dynamically attach streams to Janus that
would be even more ideal.

I think we can agree that having to dynamically restart Janus is not ideal.
Also that dynamically rewriting static config files is not ideal.

I'd like to hear your thoughts on this.

**Comment by rajmohanb** [comment]
@scandox - Yes, agree that it is not ideal to restart Janus. Also the number of video ports (network ports on which Janus listens for incoming video packets) must not be a constraint.

We need the Janus configuration file change in 2 scenarios
1. For supporting multiple camera's as in above comments
2. For issue #44 (support streaming over firefox)

Lets assume we limit max number of camera's that can be connected to the gateway to 10. Then we would need 20 streams (10x2), one each for firefox (h264) and chrome(vp8) per camera. And the identifiers would be 1..20, which is what needs to be chosen for streaming to start. So when someone clicks on a button against a camera (on the gateway dashboard), and tries to connect to the camera through the gateway, the client would need to select the appropriate id (1..20) of the camera (this has to happen without user selection). Then the issue basically comes down to associating id to each camera?

    [ IP Camera 1 ]
    type = rtp
    id = 1
    description = Camera 1 - HikVision
    audio = no
    video = yes
    ;#audioport = 5002
    ;#audiopt = 111
    ;#audiortpmap = opus/48000/2
    videoport = 5004
    videopt = 100
    videortpmap = VP8/90000
    secret = adminpwd

**Comment by scandox** [comment]
@rajmohanb yes that sounds about right. I think the details will come out in the wash.
